packages/job_execution/setup.py

```diff
@@ -55,6 +55,26 @@ def get_var(var_name):
     # In tox, it will cover them anyway.
     requirements = []
 
+DEV_RELEASE = os.environ.get("DEV_RELEASE", None) == "1"
+LOCAL_RELEASE = os.environ.get("LOCAL_RELEASE", None) == "1"
+if DEV_RELEASE:
+    _requirements = []
+    version_split = list(version.split('.'))
+    version_split[1] = str(int(version_split[1]) + 1)
+    next_version = '.'.join(version_split)
+    for requirement in requirements:
+        if requirement.startswith('galaxy-'):
+            requirement = f'{requirement}>={version},<{next_version}'
+        _requirements.append(requirement)
+    requirements = _requirements
+elif LOCAL_RELEASE:
+    _requirements = []
+    for requirement in requirements:
+        if requirement.startswith('galaxy-'):
+            requirement = f'{requirement}=={version}'
+        _requirements.append(requirement)
+    requirements = _requirements
```

Comment:
I dislike this. Can we start by just pinning everything for the meta-package like we discussed?

I was maybe okay with releasing everything all together but I don't think I'm comfortable with such aggressive version pinning. I could be talked out of my discomfort I guess though.

Reply:
I waffled on this and so I can drop the pinning for release packages. I mainly did it because the conda packages are doing it.

For local packages it's going to be necessary or you will have a very tough time getting the correct versions installed, I believe.

Reply:
> Update: I guess maybe I meant this comment for the above section. I didn't quite get what this was trying to do. I guess this local release thing is sort of specific and doesn't affect the current process?

Yes and no - the part you commented on is for exact pinning of local packages. The release packages would be pinned to "at least this exact version and less than the next major version." But as far as "the current process" goes, there isn't one since we just release versions of these packages "when someone feels like it" 😂

Reply:
That is a process and a set of standards - we have multiple projects that depend on our libraries.

Reply:
Thanks for being my therapist - it makes me sad but I guess I can deal with this.

Can we do something hacky with eval and exec though? There is not really a security concern with setup.py and the maintaince burden of this duplicated code block will be high.

Reply:
> Can we do something hacky with eval and exec though? There is not really a security concern with setup.py and the maintaince burden of this duplicated code block will be high.

I think we should be able to replace all `setup.py` files with symlinks to a general one, if we move all metadata to `setup.cfg` files, see https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html

We can probably also move the info from the `project_galaxy_*` files to the packages' `setup.cfg` files.

The "only issue" is that we already have `setup.cfg` files, which are symlinks to the main Galaxy `setup.cfg` . I think we could, as part of the Makefile, concatenate package specific (`pkg_setup.cfg`) and Galaxy `setup.cfg` in a merged `setup.cfg` .

Reply:
Thanks for the tip! This looks like a good path forward.

In the interest in keeping these PRs self-contained and reviewable maybe I'll remove the pinning `setup.py` changes from this PR and work on the `setup.cfg` conversion and any pinning we want to do in a separate PR?

Reply:
Works for me!